backend/server.js

```diff
@@ -1,29 +1,126 @@
-import bodyParser from 'body-parser'
-import cors from 'cors'
-import mongoose from 'mongoose'
+import express from 'express';
+import bodyParser from 'body-parser';
+import cors from 'cors';
+import mongoose from 'mongoose';
+import crypto from 'crypto';
+import bcrypt from 'bcrypt-nodejs';
 
-const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/authAPI"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
-mongoose.Promise = Promise
+const mongoUrl = process.env.MONGO_URL || 'mongodb://localhost/authAPI';
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
+const userSchema = new mongoose.Schema({
+  name: {
+    type: String,
+    required: [true, 'You have to enter a name.'],
+    minlength: [2, 'Oops, the entered name is too short.'],
+  },
+  password: {
+    type: String,
+    required: [true, 'You have to enter a password.'],
+    minlength: [6, 'Oops, the entered password is too short.'],
+  },
+  email: {
+    type: String,
+    required: [true, 'You have to enter an email address.'],
+    unique: true,
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex'),
+    unique: true,
+  },
+});
+
+userSchema.pre('save', async function (next) {
+  const user = this;
+
+  // Check if the password is (not) modified
+  if (!user.isModified('password')) {
+    return next();
+  }
+
+  // If the password is changed- then encrypt it
+  const salt = bcrypt.genSaltSync();
+  user.password = bcrypt.hashSync(user.password, salt);
+  next();
+});
+
+const User = mongoose.model('User', userSchema);
+
+const authenticateUser = async (req, res, next) => {
+  try {
+    const accessToken = req.header('Authorization');
+    const user = await User.findOne({ accessToken });
+    if (!user) {
+      throw 'User not found!';
+    }
+    req.user = user;
+    next();
+  } catch (err) {
+    res
+      .status(401)
+      .json({ error: 'ERROR: Something went wrong. Please try again' });
+    console.log(err);
+  }
+};
+
+// Defines the port the app will run on. Defaults to 8080, but can be
 // overridden when starting the server. For example:
 //
 //   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  res.send('Hello and welcome to Karin and Petras user API!');
+});
+
+// Signup
+app.post('/users', async (req, res) => {
+  try {
+    const { name, email, password } = req.body;
+    const user = new User({ name, email, password });
+    await user.save();
+    res
+      .status(200)
```

Comment:
I was wondering if a 201 status code is more appropriate here, as we are _creating_ and posting a new user? :)